### ISSUE TITLE
Add imported module to sys.modules

### ISSUE THREAD

**Issue by albertcthomas** [paris-saclay-cds/ramp-workflow#232]
Was there a good reason for not adding ```sys.modules[name] = module``` in the [`import_module_from_source` function](https://github.com/paris-saclay-cds/ramp-workflow/blob/master/rampwf/utils/importing.py#L9) as suggested in the [python doc](https://docs.python.org/3/library/importlib.html#importing-a-source-file-directly)?

See also issue #204 and PR #210

**Comment by albertcthomas** [comment]
This would be useful in my case and I did not have this issue when imp was used.

**Comment by agramfort** [comment]
no good reason except that it was working as is. PR welcome

>

**Comment by albertcthomas** [comment]
actually this is not completely solving my issue so I will let the function as it is for now

**Comment by albertcthomas** [comment]
Coming back to this after further investigation and a better understanding of what happens when we want to pickle dynamically imported objects.

If we don't put the name of the dynamically imported module in `sys.modules`, then I think, although I am not completely sure about this, that when we want to pickle an object of this dynamically imported module with `cloudpickle`, `cloudpickle` sees it as a dynamically imported object and knows how to pickle these kinds of objects and it basically just works. If we use `pickle` instead `cloudpickle` then it does not work because `pickle` does not handle these situations.

If we put the name of the module in `sys.modules` then `cloudpickle` and `pickle` do not work even though in theory they should know how to import the modules containing the objects they want to pickle. The reason for this is that when you run `ramp-test` you don't have the root directory (the one containing `problem.py`, `data/` and `submissions/` and where we should run `ramp-test`, in `sys.path` which is expected when you define a `console_script` entry point in `setup.py`. I have to say that when I discovered this I was suprised at first because if you had a `ramp-test.py` file in the root dir and ran `python ramp-test.py` then the current directory would be added to `sys.path`. If you add the root dir to `sys.path` then both `cloudpickle` and `pickle` work. Note that since the introduction of [Implicit namespace packages](https://www.python.org/dev/peps/pep-0420/) in python 3.3 we don't need `__init__` files anymore in subdirectories.

When `imp.load_source` was used, the name of the module was put in `sys.modules` and when I checkout the commit before the removal of imp in favor of importlib (see PR #210), `cloudpickle` was not working.

Also if you add the name to `sys.modules` and try to pickle with `ramp-test --pickle` in the new `iris` rampkit using the scikit-learn pipeline workflow then this works because the module of the submission estimator is a `sklearn` module which is in `sys.path`. This would not be the case with a custom estimator.

Finally we had a case where we had to use `dill` to pickle a submission instead of `cloudpickle` which was not working (see cloudpipe/cloudpickle#381) . But to make `dill` work I had to add the name of the imported submission module to `sys.modules` (in this use case, the root dir was in `sys.path`).

**Comment by albertcthomas** [comment]
Interestingly when you run a tool like `pytest`, depending on your configuration, `pytest` can modify `sys.path` to import the test modules. However there is now an `importlib` parameter that you can turn on which does not modify `sys.path` and do not put the module in `sys.modules` (see the corresponding [pytest code](https://github.com/pytest-dev/pytest/blob/master/src/_pytest/pathlib.py#L442))

**Comment by kegl** [comment]
My head just exploded, brain tissue on the screen :)

**Comment by albertcthomas** [comment]
TLDR, there is a good reason for not putting the imported module in `sys.modules`, if you do it you break `cloudpickle`.

**Comment by glemaitre** [comment]
I absolutely do not remember TBH. I think that I change the import first in `ramp-board` and maybe I encounter such issue there that made me remove the `sys.modules` but I don't remember.

In the end, are you still having an issue to pickle using cloudpickle when you have your own estimator?

**Comment by albertcthomas** [comment]
Cloudpickle is working for all our estimators except one of them which is based on gpytorch. For this one we use dill and we need to put the module name in sys.modules. I think we can live with that for now. Cloudpickle devs might investigate if they can pickle it with cloudpickle.
